language/move-vm/types/src/values/values_impl.rs

```diff
@@ -1988,66 +1973,57 @@ impl<'a, 'b> serde::Serialize for AnnotatedValue<'a, 'b, Type, ValueImpl> {
             (Type::Vector(ty), ValueImpl::Container(r)) => {
                 let ty = &**ty;
                 match (ty, &*r.borrow()) {
-                    (Type::Vector(_), Container::General(v))
-                    | (Type::Struct(_), Container::General(v))
-                    | (Type::Address, Container::General(v)) => {
+                    (Type::U8, Container::U8(v)) => v.serialize(serializer),
+                    (Type::U64, Container::U64(v)) => v.serialize(serializer),
+                    (Type::U128, Container::U128(v)) => v.serialize(serializer),
+                    (Type::Bool, Container::Bool(v)) => v.serialize(serializer),
+
+                    (_, Container::General(v)) => {
                         let mut t = serializer.serialize_seq(Some(v.len()))?;
                         for val in v {
                             t.serialize_element(&AnnotatedValue { ty, val })?;
                         }
                         t.end()
                     }
 
-                    (Type::U8, Container::U8(v)) => serialize_vec!(U8, ty, v),
-                    (Type::U64, Container::U64(v)) => serialize_vec!(U64, ty, v),
-                    (Type::U128, Container::U128(v)) => serialize_vec!(U128, ty, v),
-                    (Type::Bool, Container::Bool(v)) => serialize_vec!(Bool, ty, v),
-
-                    (ty, container) => Err(S::Error::custom(
-                        VMStatus::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR).with_message(
-                            format!("cannot serialize container {:?} as {:?}", container, ty),
-                        ),
-                    )),
+                    (ty, container) => Err(invariant_violation::<S>(format!(
```

Comment:
In a follow-up, we should probably add explicit pattern matching here and elsewhere to future-proof this...